### Matthew Riddell 
### D00245674
### MSc in Data Analytics
### Dissertation
### Initial Data Exploration Jupyter Notebook

In this part, I will conduct Initial EDA of the scraped ECI petition datasets and the imported Worldstat dataset

For this part I will investigate the data itself i.e. dataset size and shape, the variable types, missing values, duplicates etc so that I can plan on how to properly merge the datasets and clean them for proper Data Analysis 

Data Exploration conducted before Data Preparation and Cleaning

### Data Description

The extracted variables from each The Petition Datasets:

- Country Name
- Number of Signatures
- Country Minimum Threshold
- Percentage of Threshold

These variables from the Worldstat Dataset will be selected:

- Country Name
- Population
- GDP
- Unemployment Rate
- Life Expectancy
- Population Density
- Gross Tertiary Education Enrollment
- Urban Population
- Labor Participation

### Link to each Dataset:

### ECI Petitions:
https://citizens-initiative.europa.eu/initiatives/details/2024/000001_en ECI 01: Ban on conversion practices in the European Union

https://citizens-initiative.europa.eu/initiatives/details/2024/000002_en ECI 02: European Citizens' Initiative in Defence of Agriculture and Rural Economy in Europe

https://citizens-initiative.europa.eu/initiatives/details/2024/000004_en ECI 04: My Voice, My Choice: For Safe And Accessible Abortion

https://citizens-initiative.europa.eu/initiatives/details/2024/000006_en ECI 06: Air-Quotas

https://citizens-initiative.europa.eu/initiatives/details/2024/000007_en ECI 07: Stop Destroying Videogames

https://citizens-initiative.europa.eu/initiatives/details/2024/000008_en ECI 08: Stop Cruelty Stop Slaughter

https://citizens-initiative.europa.eu/initiatives/details/2024/000009_en ECI 09: STOP FAKE FOOD: ORIGIN ON LABEL

### Worldstat Kaggle Dataset
https://www.kaggle.com/datasets/nelgiriyewithana/countries-of-the-world-2023

### Mahara Portfolio:
https://mahara.dkit.ie/view/view.php?t=f816cb6da6dd833aaa2f

### Github Repository:
https://github.com/Matthew-Riddell/MSc-Dissertation

In [1]:
import pandas as pd 
import numpy as np

#### File path and file loading

In [2]:
file_path = "C:/Users/Matty/Documents/College Notes & Assignments/Year 5/Dissertation/MSc-Dissertation/data/Unprepared Datasets/"

eci_01 = pd.read_csv(file_path + "eci_table_01.csv")
eci_02 = pd.read_csv(file_path + "eci_table_02.csv")
eci_04 = pd.read_csv(file_path + "eci_table_04.csv")
eci_06 = pd.read_csv(file_path + "eci_table_06.csv")
eci_07 = pd.read_csv(file_path + "eci_table_07.csv")
eci_08 = pd.read_csv(file_path + "eci_table_08.csv")
eci_09 = pd.read_csv(file_path + "eci_table_09.csv")

worldstat = pd.read_csv(file_path + "world-data-2023.csv")

### Dataset Shape

In [3]:
print("ECI 01:", eci_01.shape)
print("ECI 02:", eci_02.shape)
print("ECI 04:", eci_04.shape)
print("ECI 06:", eci_06.shape)
print("ECI 07:", eci_07.shape)
print("ECI 08:", eci_08.shape)
print("ECI 09:", eci_09.shape)
print("Worldstat:", worldstat.shape)

ECI 01: (28, 4)
ECI 02: (28, 4)
ECI 04: (28, 4)
ECI 06: (28, 4)
ECI 07: (28, 4)
ECI 08: (28, 4)
ECI 09: (28, 4)
Worldstat: (195, 35)


All 7 petitions contain the same overall structure with 28 rows each and 4 columns 

The worldstat dataset contains a much more complex structure with 195 rows and 35 columns

#### Display first 5 rows of each dataset

In [5]:
eci_01.head()

,Country,Signatures,Threshold,Percentage
0,Austria,"10,724","13,395",80.06%
1,Belgium,"47,459","14,805",320.56%
2,Bulgaria,"2,386","11,985",19.91%
3,Croatia,"8,574","8,460",101.35%
4,Cyprus,696,"4,230",16.45%


In [6]:
eci_02.head()

,Country,Signatures,Threshold,Percentage
0,Austria,109,"13,395",0.81%
1,Belgium,160,"14,805",1.08%
2,Bulgaria,47,"11,985",0.39%
3,Croatia,76,"8,460",0.90%
4,Cyprus,27,"4,230",0.64%


In [7]:
eci_04.head()

,Country,Signatures,Threshold,Percentage
0,Austria,"25,373","13,395",189.42%
1,Belgium,"18,313","14,805",123.69%
2,Bulgaria,"6,552","11,985",54.67%
3,Croatia,"65,963","8,460",779.70%
4,Cyprus,"3,228","4,230",76.31%


In [8]:
eci_06.head()

,Country,Signatures,Threshold,Percentage
0,Austria,156,"13,395",1.16%
1,Belgium,257,"14,805",1.74%
2,Bulgaria,17,"11,985",0.14%
3,Croatia,24,"8,460",0.28%
4,Cyprus,8,"4,230",0.19%


In [9]:
eci_07.head()

,Country,Signatures,Threshold,Percentage
0,Austria,"20,714","13,395",154.64%
1,Belgium,"31,846","14,805",215.10%
2,Bulgaria,"14,238","11,985",118.80%
3,Croatia,"14,403","8,460",170.25%
4,Cyprus,"1,997","4,230",47.21%


In [10]:
eci_08.head()

,Country,Signatures,Threshold,Percentage
0,Austria,714,"13,395",5.33%
1,Belgium,564,"14,805",3.81%
2,Bulgaria,122,"11,985",1.02%
3,Croatia,472,"8,460",5.58%
4,Cyprus,29,"4,230",0.69%


In [11]:
eci_09.head()

,Country,Signatures,Threshold,Percentage
0,Austria,352,"13,395",2.63%
1,Belgium,426,"14,805",2.88%
2,Bulgaria,185,"11,985",1.54%
3,Croatia,217,"8,460",2.57%
4,Cyprus,63,"4,230",1.49%


In [ ]:
worldstat.head()

,Country,Density\n(P/Km2),Abbreviation,Agricultural Land( %),Land Area(Km2),Armed Forces size,Birth Rate,Calling Code,Capital/Major City,Co2-Emissions,...,Out of pocket health expenditure,Physicians per thousand,Population,Population: Labor force participation (%),Tax revenue (%),Total tax rate,Unemployment rate,Urban_population,Latitude,Longitude
0,Afghanistan,60,AF,58.10%,"652,230","323,000",32.49,93.0,Kabul,"8,672",...,78.40%,0.28,"38,041,754",48.90%,9.30%,71.40%,11.12%,"9,797,273",33.939110,67.709953
1,Albania,105,AL,43.10%,"28,748","9,000",11.78,355.0,Tirana,"4,536",...,56.90%,1.20,"2,854,191",55.70%,18.60%,36.60%,12.33%,"1,747,593",41.153332,20.168331
2,Algeria,18,DZ,17.40%,"2,381,741","317,000",24.28,213.0,Algiers,"150,006",...,28.10%,1.72,"43,053,054",41.20%,37.20%,66.10%,11.70%,"31,510,100",28.033886,1.659626
3,Andorra,164,AD,40.00%,468,NaN,7.20,376.0,Andorra la Vella,469,...,36.40%,3.33,"77,142",NaN,NaN,NaN,NaN,"67,873",42.506285,1.521801
4,Angola,26,AO,47.50%,"1,246,700","117,000",40.73,244.0,Luanda,"34,693",...,33.40%,0.21,"31,825,295",77.50%,9.20%,49.10%,6.89%,"21,061,025",-11.202692,17.873887


In [12]:
print("First 5 rows of ECI 01 data:")
print(eci_01.head())

First 5 rows of ECI 01 data:
    Country Signatures Threshold Percentage
0   Austria     10,724    13,395     80.06%
1   Belgium     47,459    14,805    320.56%
2  Bulgaria      2,386    11,985     19.91%
3   Croatia      8,574     8,460    101.35%
4    Cyprus        696     4,230     16.45%


In [13]:
print("First 5 rows of ECI 02 data:")
print(eci_02.head())

First 5 rows of ECI 02 data:
    Country Signatures Threshold Percentage
0   Austria        109    13,395      0.81%
1   Belgium        160    14,805      1.08%
2  Bulgaria         47    11,985      0.39%
3   Croatia         76     8,460      0.90%
4    Cyprus         27     4,230      0.64%


In [14]:
print("First 5 rows of ECI 04 data:")
print(eci_04.head())

First 5 rows of ECI 04 data:
    Country Signatures Threshold Percentage
0   Austria     25,373    13,395    189.42%
1   Belgium     18,313    14,805    123.69%
2  Bulgaria      6,552    11,985     54.67%
3   Croatia     65,963     8,460    779.70%
4    Cyprus      3,228     4,230     76.31%


In [15]:
print("First 5 rows of ECI 06 data:")
print(eci_06.head())

First 5 rows of ECI 06 data:
    Country Signatures Threshold Percentage
0   Austria        156    13,395      1.16%
1   Belgium        257    14,805      1.74%
2  Bulgaria         17    11,985      0.14%
3   Croatia         24     8,460      0.28%
4    Cyprus          8     4,230      0.19%


In [16]:
print("First 5 rows of ECI 07 data:")
print(eci_07.head())

First 5 rows of ECI 07 data:
    Country Signatures Threshold Percentage
0   Austria     20,714    13,395    154.64%
1   Belgium     31,846    14,805    215.10%
2  Bulgaria     14,238    11,985    118.80%
3   Croatia     14,403     8,460    170.25%
4    Cyprus      1,997     4,230     47.21%


In [17]:
print("First 5 rows of ECI 08 data:")
print(eci_08.head())

First 5 rows of ECI 08 data:
    Country Signatures Threshold Percentage
0   Austria        714    13,395      5.33%
1   Belgium        564    14,805      3.81%
2  Bulgaria        122    11,985      1.02%
3   Croatia        472     8,460      5.58%
4    Cyprus         29     4,230      0.69%


In [18]:
print("First 5 rows of ECI 09 data:")
print(eci_09.head())

First 5 rows of ECI 09 data:
    Country Signatures Threshold Percentage
0   Austria        352    13,395      2.63%
1   Belgium        426    14,805      2.88%
2  Bulgaria        185    11,985      1.54%
3   Croatia        217     8,460      2.57%
4    Cyprus         63     4,230      1.49%


In [19]:
print("First 5 rows of the World Stat dataset:")
print(worldstat.head())

First 5 rows of the World Stat dataset:
       Country Density\n(P/Km2) Abbreviation Agricultural Land( %)  \
0  Afghanistan               60           AF                58.10%   
1      Albania              105           AL                43.10%   
2      Algeria               18           DZ                17.40%   
3      Andorra              164           AD                40.00%   
4       Angola               26           AO                47.50%   

  Land Area(Km2) Armed Forces size  Birth Rate  Calling Code  \
0        652,230           323,000       32.49          93.0   
1         28,748             9,000       11.78         355.0   
2      2,381,741           317,000       24.28         213.0   
3            468               NaN        7.20         376.0   
4      1,246,700           117,000       40.73         244.0   

  Capital/Major City Co2-Emissions  ... Out of pocket health expenditure  \
0              Kabul         8,672  ...                           78.40%   
1 

All 7 petitions contain the same overall structure with 28 rows each, one for each european country + the total signatures row, with 4 columns (country name, signatures, threshold, percentage)

The worldstat dataset contains a much more complex structure with 195 rows, one for each world country, and 35 fields containing different country level statistics

#### Check basic information about the dataset

In [20]:
print("Information about the ECI 01 dataset:")
print(eci_01.info())

Information about the ECI 01 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [21]:
print("Information about the ECI 02 dataset:")
print(eci_02.info())

Information about the ECI 02 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [22]:
print("Information about the ECI 04 dataset:")
print(eci_04.info())

Information about the ECI 04 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [23]:
print("Information about the ECI 06 dataset:")
print(eci_06.info())

Information about the ECI 06 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [24]:
print("Information about the ECI 07 dataset:")
print(eci_07.info())

Information about the ECI 07 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [25]:
print("Information about the ECI 08 dataset:")
print(eci_08.info())

Information about the ECI 08 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [26]:
print("Information about the ECI 09 dataset:")
print(eci_09.info())

Information about the ECI 09 dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Country     28 non-null     object
 1   Signatures  28 non-null     object
 2   Threshold   27 non-null     object
 3   Percentage  27 non-null     object
dtypes: object(4)
memory usage: 1.0+ KB
None


In [27]:
print("Information about the Worldstat dataset:")
print(worldstat.info())

Information about the Worldstat dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 35 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   Country                                    195 non-null    object 
 1   Density
(P/Km2)                            195 non-null    object 
 2   Abbreviation                               188 non-null    object 
 3   Agricultural Land( %)                      188 non-null    object 
 4   Land Area(Km2)                             194 non-null    object 
 5   Armed Forces size                          171 non-null    object 
 6   Birth Rate                                 189 non-null    float64
 7   Calling Code                               194 non-null    float64
 8   Capital/Major City                         192 non-null    object 
 9   Co2-Emissions                              188 non-null  

Most of the numerical variables have been imported as object data types 

due to them containing commas, percentage symbols and other strings within both the scraped petition datasets and the imported worldstat dataset

### Missing values per column

In [30]:
eci_01.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [31]:
eci_02.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [32]:
eci_04.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [33]:
eci_06.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [34]:
eci_07.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [35]:
eci_08.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [36]:
eci_09.isnull().sum()

Country       0
Signatures    0
Threshold     1
Percentage    1
dtype: int64

In [29]:
worldstat.isnull().sum()

Country                                       0
Density\n(P/Km2)                              0
Abbreviation                                  7
Agricultural Land( %)                         7
Land Area(Km2)                                1
Armed Forces size                            24
Birth Rate                                    6
Calling Code                                  1
Capital/Major City                            3
Co2-Emissions                                 7
CPI                                          17
CPI Change (%)                               16
Currency-Code                                15
Fertility Rate                                7
Forested Area (%)                             7
Gasoline Price                               20
GDP                                           2
Gross primary education enrollment (%)        7
Gross tertiary education enrollment (%)      12
Infant mortality                              6
Largest city                            

each petition contains a missing value in threshold and percentage

this is due to the total signatures row which isnt relevent to the country level data

the world stat datat contains a lot of missing values scattered across almost every column except for country and population density

### Total missing values

In [45]:
total_missing = eci_01.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [46]:
total_missing = eci_02.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [47]:
total_missing = eci_04.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [48]:
total_missing = eci_06.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [49]:
total_missing = eci_07.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [50]:
total_missing = eci_08.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [51]:
total_missing = eci_09.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 2


In [52]:
total_missing = worldstat.isna().sum().sum()
print("Total missing values in the dataset:", total_missing)

Total missing values in the dataset: 341


each petition contains a missing value in threshold and percentage

this is due to the total signatures row which isnt relevent to the country level data

the world stat datat contains a lot of missing values scattered across almost every column except for country and population density

total missing values per petition: 2

total missing values in worldstat dataset: 341

### Display column names

In [38]:
print("ECI 01 column names:")
print(eci_01.columns)

ECI 01 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [39]:
print("ECI 02 column names:")
print(eci_02.columns)

ECI 02 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [40]:
print("ECI 04 column names:")
print(eci_04.columns)

ECI 04 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [41]:
print("ECI 06 column names:")
print(eci_06.columns)

ECI 06 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [42]:
print("ECI 07 column names:")
print(eci_07.columns)

ECI 07 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [43]:
print("ECI 08 column names:")
print(eci_08.columns)

ECI 08 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [44]:
print("ECI 09 column names:")
print(eci_09.columns)

ECI 09 column names:
Index(['Country', 'Signatures', 'Threshold', 'Percentage'], dtype='object')


In [37]:
print("Worldstat column names:")
print(worldstat.columns)

Worldstat column names:
Index(['Country', 'Density\n(P/Km2)', 'Abbreviation', 'Agricultural Land( %)',
       'Land Area(Km2)', 'Armed Forces size', 'Birth Rate', 'Calling Code',
       'Capital/Major City', 'Co2-Emissions', 'CPI', 'CPI Change (%)',
       'Currency-Code', 'Fertility Rate', 'Forested Area (%)',
       'Gasoline Price', 'GDP', 'Gross primary education enrollment (%)',
       'Gross tertiary education enrollment (%)', 'Infant mortality',
       'Largest city', 'Life expectancy', 'Maternal mortality ratio',
       'Minimum wage', 'Official language', 'Out of pocket health expenditure',
       'Physicians per thousand', 'Population',
       'Population: Labor force participation (%)', 'Tax revenue (%)',
       'Total tax rate', 'Unemployment rate', 'Urban_population', 'Latitude',
       'Longitude'],
      dtype='object')


The column structure is consistent across the petitions, this will allow for easy combination

the world stat structure is a little more complex and will require careful merging based on the countries in the petition datasets and a petition id 

### Duplicate checking

In [57]:
print("ECI 01:", eci_01.duplicated().sum())
print("ECI 02:", eci_02.duplicated().sum())
print("ECI 04:", eci_04.duplicated().sum())
print("ECI 06:", eci_06.duplicated().sum())
print("ECI 07:", eci_07.duplicated().sum())
print("ECI 08:", eci_08.duplicated().sum())
print("ECI 09:", eci_09.duplicated().sum())
print("Worldstat:", worldstat.duplicated().sum())

ECI 01: 0
ECI 02: 0
ECI 04: 0
ECI 06: 0
ECI 07: 0
ECI 08: 0
ECI 09: 0
Worldstat: 0


Each dataset contains 0 duplicates

### Initial EDA Observations:

The ECI datasets contain petition data and summary rows that will require cleaning before proper analysis

most variables are stored as text rather than numerical values and will require cleaning and transforming into numericals before statistical analysis

missing values exist in the Worldstat dataset and will require investigation during preparation

country naming conventions may differ between datasets and will need to be standardised before dataset merging

Due to the nature of the uncleaned, unprepared data

Proper Data exploration cannot be conducted until the data has been properly cleaned and prepared

due to most numerical fields containing string data such as commas, and percentage signs

as well as missing values

### Further and extensive Data exploration will be conducted after Data Cleaning and Preparation